src/celeritas/phys/PhysicsStepUtils.hh

```diff
@@ -197,17 +200,19 @@ inline CELER_FUNCTION ParticleTrackView::Energy
         auto grid_id = physics.value_grid(VGT::range, ppid);
         CELER_ASSERT(grid_id);
 
-        // Recalculate beginning-of-step range (instead of storing)
-        auto calc_range = physics.make_calculator<RangeCalculator>(grid_id);
-        real_type range = calc_range(Energy{pre_step_energy});
+        // Use the range limit stored from calc_physics_step_limit
+        real_type range = physics.dedx_range();
         if (step == range)
         {
             // TODO: eloss should be pre_step_energy at this point only if the
             // range was the step limiter (step == range), and if the
             // range-to-step conversion was 1.
             return Energy{pre_step_energy};
         }
-        CELER_ASSERT(range > step);
+        step = celeritas::min<real_type>(range, step);
+        // TODO: replace this temporary tweak with CELER_ASSERT(range > step)
+        // when numerical instability (off by 1 ulp) of the conversion from
+        // range to step is fixed.
```

Comment:
Can you move the `step` line here to the last line of `PhysicsTrackView::range_to_step`? Basically, the bug is that `range_to_step` is slightly *increasing* the returned step from what should be possible: so that function should be responsible for ensuring it's within the acceptable limits. When that's done, this ASSERT can be left in place, and the MSC assertion can probably drop the `|| soft_equiv`.

Reply:
You are absolutely right!